### ISSUE TITLE
Lots of parked "Fiber thread for block at" threads

### ISSUE THREAD

**Issue by Unknown-Guy** [jruby/jruby#6004]
**Environment**
 
* `JRUBY_OPTS='--dev --headless -J-XX:+UseStringDeduplication'`
* `jruby 9.2.9.0 (2.5.7) 2019-10-30 458ad3e OpenJDK 64-Bit Server VM 25.212-b03 on 1.8.0_212-b03 [darwin-x86_64]`
* `Darwin Me-MBP-3.local 19.2.0 Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64 x86_64`

Other relevant info you may wish to add:
*  does not happen on 9.2.7.0, same issue on 9.2.8.0
* [thread dump](https://gist.github.com/Unknown-Guy/36dad4b61e53b5366b49208baafd9283) 

**Expected Behavior**

Does not generate excessive amount of threads

**Actual Behavior**

VisualVM shows lots of threads in for `Fiber thread for block at: uri:classloader:/jruby/kernel/enumerator.rb:62` in `Park` state and running our apps full test suite ends up with running out of threads at around 4k.

**Comment by headius** [comment]
So something changed in 9.2.8 that's causing these to not get cleaned up.

Fibers are used to implement `Enumerator#next`, and on JRuby each fiber requires a native thread. If you do not `next` your way completely through the collection, that means the fiber thread is sitting there waiting for another call. We have taken pains to make sure that fibers that are abandoned will eventually get cleaned up (usually via GC triggering finalization) but perhaps something broke along the way that's causing this to fail now.

**Comment by headius** [comment]
Ok, so the big thing that landed in 9.2.8.0 was the actual move to a pure-Ruby `Enumerator#next` based on Fiber. This was done to eliminate a second "fiber-like" bit of logic we had been maintaining separately.

So we need to figure out why these fibers are sticking around.

1. Do you know what code is causing them to be created? It would be `Enumerator#next` logic, clearly. You may be able to pass `-Xjit.threshold=0` to JRuby to force more code to JIT compile and show up in the stack.
2. Does forcing GC clean them up? You can connect with `jconsole` or `jvisualvm` and force a GC (or two or three) and see if they clear up.
3. Finally, we'll want to try to get a heap dump when there are a large number of these fibers hanging around. That will let us track why they're being held in memory, when they should GC and go away. You can also do this with `jconsole` or `jvisualvm`. Zip up the resulting file and post it online for us to download. NOTE: It will contain most of the data your app has in-memory, so if that is sensitive we'll discuss how to securely get a copy to us to inspect.

Thank you for reporting this! I worked through a few other issues with these fibers not being collected and thought I got them all. It seems there may still be something left.

**Comment by Unknown-Guy** [comment]
Thanks, will do some digging on Monday. And no, didn't find any obvious code that would cause this.

**Comment by Unknown-Guy** [comment]
Force GC does nothing.

**Comment by headius** [comment]
@Unknown-Guy Thank you! This will help us investigate.

**Comment by headius** [comment]
I have found a possible culprit.

Based on the heap dump, most of the living Enumerator (RubyEnumerator) objects appear to be alive because they're referenced from a fiber's thread stack. For enumerators that wrap another object, this should not be the case; the fiber would only reference that object (e.g. to call each inside the fiber) and the enumerator itself could properly GC. However the stuck enumerators all appear to be initiated via the `Enumerator::Lazy` constructor [here](https://github.com/jruby/jruby/blob/master/core/src/main/ruby/jruby/kernel/enumerator.rb#L118).

This logic evolved separate from the fiber-based `Enumerator#next` logic, and it appears to defeat the logic I added in 6c232abbda5 to avoid any hard references from the fiber to the associated enumerator. It appears that the reference chain goes backward from the enumerator object to the block in `Enumerator::Lazy#initialize`, which is held by a `Enumerator::Generator`, which is the object referenced by the fiber in `Enumerator::FiberGenerator`. So in a round-about way, the enumerator is held by a block, which is passed to a generator, which becomes the "object" that the fiber calls `each` against.

It's likely that we will need to mimic the old `Enumerator#next` logic which used weak references to ensure the enumerator was never hard referenced by the fiber thread (assuming the old logic really doesn't have this bug).

**Comment by headius** [comment]
The script I used to confirm this leaks is provided here:

```ruby
while true
  Enumerator::Lazy.new([1,2,3]) {|x| x.yield(1) }.next
  JRuby.gc
end
```

This script rapidly spins up threads that do not go away.

If the `Enumerator::Lazy` construction is replaced with `[1,2,3].each.next` or `[1,2,3].lazy.each.next` this runs fine.

**Comment by headius** [comment]
Hmm, I may have found a bug that also exists in JRuby versions prior to the fiber-based `Enumerator#next` logic. The script above also blows up in JRuby 9.2.7.0 and 9.1.17.0.

Could you do a couple things for me please?

1. Audit your code and dependencies for any instances of `Lazy.new` similar to what I have above.
2. Confirm for me that nothing else changed in your app while upgrading to JRuby 9.2.9.0.

I may need you to provide a new heap dump plus thread dump, so I can see the full call stack of each thread (I might be able to get that out of your previous dump, but I'm not sure).

**Comment by headius** [comment]
I had forgotten that several `Lazy` methods themselves call `Lazy.new`, such as for `Lazy#reject`:

https://github.com/jruby/jruby/blob/a50171981f330a4697d83a8abdb1527f8992a60e/core/src/main/ruby/jruby/kernel/enumerator.rb#L231-L237

This lazy methods appear to leak as well when used with `Enumerator#next`, for the same reasons.

**Comment by headius** [comment]
This change appears to help the `Lazy` leak. I wonder if it will help your environment? It could be that the upgrade JRuby also uses `Lazy` more, perhaps due to a standard library update.

```diff
diff --git a/core/src/main/ruby/jruby/kernel/enumerator.rb b/core/src/main/ruby/jruby/kernel/enumerator.rb
index 21dcb564b9..8d75367ad7 100644
--- a/core/src/main/ruby/jruby/kernel/enumerator.rb
+++ b/core/src/main/ruby/jruby/kernel/enumerator.rb
@@ -115,10 +115,14 @@ class Enumerator
 
   class Lazy < Enumerator
 
-    def initialize(obj, size = nil)
-      _block_error(:new) unless block_given?
+    def initialize(obj, size = nil, &block)
+      _block_error(:new) unless block
       @receiver = obj
-      super(size) do |yielder, *args|
+      super(size, &self.class.make_proc(obj, &block))
+    end
+
+    def self.make_proc(obj)
+      proc do |yielder, *args|
         catch yielder do
           obj.each(*args) do |*x|
             yield yielder, *x
```

If it looks ok running tests locally I'll push this and a test for it.

**Comment by Unknown-Guy** [comment]
Didn't find anything interesting in our codebase, but my guess was that something in capybara does this. This seems to be one of the issues: https://github.com/teamcapybara/capybara/blob/2e052792a59d0c183ff42009cc73bf7e9eae43c6/lib/capybara/result.rb#L153

Changed the line to compare with non-existing `9.2.10.0` and now thread count went down from 177 live and 194 peak to 69/80

**Comment by Unknown-Guy** [comment]
Tried adding your patch just before `environment.rb` before loading rails - did not help.

**Comment by headius** [comment]
@Unknown-Guy That's a good discovery! Would you say it runs well with that line modified for 9.2.10.0? If so that certainly points at lazy being a likely culprit. It also explains why this would have worked a few versions ago and started failing recently.

There's some tweakiness about getting changes to the "kernel" Ruby files to get picked up by a runtime, so I'll push the patch and have you try again.

**Comment by headius** [comment]
I have pushed #6021 with this first fix, give that branch a try and let me know if the situation has improved at all.

I wish I could reproduce locally, but without that the next step will be getting a new heap dump + thread dump so I can correlate the live threads with all references they are keeping in memory.

**Comment by Unknown-Guy** [comment]
Unfortunately don't see any improvement. Uploaded dumps [here](https://failiem.lv/u/hqyh9s8w).

**Comment by boris-petrov** [comment]
@headius - I just noticed the same issue. Again with Capybara. Latest version of everything. Is there any workaround? Do you have ideas where the problem is? And is it going to be fixed? Just asking so I can decide what to do. Thanks!

**Comment by headius** [comment]
@boris-petrov Some parts of this were fixed, but definitely not all of them. There are some Fiber patterns that are going to be very difficult to fix, since they keep threads alive with no iindication that they won't be used again (other than eventually getting garbage collected, which may take a long time).

If it's possible, could you get a new thread dump of the issue you're seeing? Ideally, running with JRuby flags `-Xjit.threshold=0` (to make sure we get all the Ruby lines into the Java trace) and `-Xbacktrace.style=full`.

I'll keep this in my queue and try to revisit some of these cases, but there may be no way to fix all of the problems in our Enumerable::Lazy implementation.

**Comment by boris-petrov** [comment]
@headius - thanks for the support.

I did a thread dump around the end of our integration tests. I hope I've added correctly these flags that you mentioned, please tell me if I haven't. I won't attach the whole file because it's 28MB uncompressed (if you need it, I'll add it of course). There are 3232 occurrences of "Fiber thread for". The stacktrace is (I think for all of them):

```
"Fiber thread for block at: uri:classloader:/jruby/kernel/enumerator.rb:62" #236 daemon prio=5 os_prio=0 cpu=321.43ms elapsed=1378.86s tid=0x00007f58cc39d800 nid=0x2c442 waiting on condition  [0x00007f58a55d1000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@13.0.2/Native Method)
	- parking to wait for  <0x0000000728012830> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(java.base@13.0.2/LockSupport.java:194)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@13.0.2/AbstractQueuedSynchronizer.java:2081)
	at java.util.concurrent.ArrayBlockingQueue.take(java.base@13.0.2/ArrayBlockingQueue.java:417)
	at org.jruby.ext.fiber.FiberQueue$1.run(FiberQueue.java:55)
	at org.jruby.ext.fiber.FiberQueue$1.run(FiberQueue.java:52)
	at org.jruby.RubyThread.executeTask(RubyThread.java:1585)
	at org.jruby.RubyThread.executeTask(RubyThread.java:1571)
	at org.jruby.ext.fiber.FiberQueue.pop(FiberQueue.java:129)
	at org.jruby.ext.fiber.FiberQueue.pop(FiberQueue.java:107)
	at org.jruby.ext.fiber.ThreadFiber.exchangeWithFiber(ThreadFiber.java:114)
	at org.jruby.ext.fiber.ThreadFiber.yield(ThreadFiber.java:220)
	at org.jruby.ext.fiber.ThreadFiber.yield(ThreadFiber.java:227)
	at org.jruby.ext.fiber.ThreadFiber$INVOKER$s$yield.call(ThreadFiber$INVOKER$s$yield.gen)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:74)
	at uri_3a_classloader_3a_.jruby.kernel.enumerator.invokeOther0:yield(uri:classloader:/jruby/kernel/enumerator.rb:65)
	at uri_3a_classloader_3a_.jruby.kernel.enumerator.RUBY$block$initialize$2(uri:classloader:/jruby/kernel/enumerator.rb:65)
	at java.lang.invoke.DirectMethodHandle$Holder.invokeStatic(java.base@13.0.2/DirectMethodHandle$Holder)
	at java.lang.invoke.LambdaForm$MH/0x00000008005b3840.invoke(java.base@13.0.2/LambdaForm$MH)
	at java.lang.invoke.Invokers$Holder.invokeExact_MT(java.base@13.0.2/Invokers$Holder)
	at org.jruby.runtime.CompiledIRBlockBody.callDirect(CompiledIRBlockBody.java:138)
	at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58)
	at org.jruby.runtime.Block.call(Block.java:143)
	at org.jruby.RubyYielder$BlockCallbackImpl.call(RubyYielder.java:93)
	at org.jruby.runtime.CallBlock19.call(CallBlock19.java:68)
	at org.jruby.runtime.Block.call(Block.java:139)
	at org.jruby.RubyYielder.yield(RubyYielder.java:118)
	at org.jruby.RubyYielder$INVOKER$i$0$0$yield.call(RubyYielder$INVOKER$i$0$0$yield.gen)
	at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodN.call(JavaMethod.java:815)
	at org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:200)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:172)
	at uri_3a_classloader_3a_.jruby.kernel.enumerator.invokeOther3:yield(uri:classloader:/jruby/kernel/enumerator.rb:229)
	at uri_3a_classloader_3a_.jruby.kernel.enumerator.RUBY$block$select$1(uri:classloader:/jruby/kernel/enumerator.rb:229)
	at java.lang.invoke.DirectMethodHandle$Holder.invokeStatic(java.base@13.0.2/DirectMethodHandle$Holder)
	at java.lang.invoke.LambdaForm$MH/0x0000000801b10440.invoke(java.base@13.0.2/LambdaForm$MH)
	at java.lang.invoke.DelegatingMethodHandle$Holder.delegate(java.base@13.0.2/DelegatingMethodHandle$Holder)
	at java.lang.invoke.LambdaForm$MH/0x00000008007c3840.guard(java.base@13.0.2/LambdaForm$MH)
	at java.lang.invoke.Invokers$Holder.linkToCallSite(java.base@13.0.2/Invokers$Holder)
	at uri_3a_classloader_3a_.jruby.kernel.enumerator.RUBY$block$make_proc$3(uri:classloader:/jruby/kernel/enumerator.rb:128)
	at java.lang.invoke.DirectMethodHandle$Holder.invokeStatic(java.base@13.0.2/DirectMethodHandle$Holder)
	at java.lang.invoke.LambdaForm$MH/0x00000008005b3c40.invoke(java.base@13.0.2/LambdaForm$MH)
	at java.lang.invoke.Invokers$Holder.invokeExact_MT(java.base@13.0.2/Invokers$Holder)
	at org.jruby.runtime.CompiledIRBlockBody.yieldDirect(CompiledIRBlockBody.java:148)
	at org.jruby.runtime.BlockBody.yield(BlockBody.java:114)
	at org.jruby.runtime.Block.yieldValues(Block.java:200)
	at org.jruby.RubyEnumerator$2.call(RubyEnumerator.java:404)
	at org.jruby.runtime.BlockCallback.call(BlockCallback.java:40)
	at org.jruby.runtime.CallBlock.doYield(CallBlock.java:96)
	at org.jruby.runtime.BlockBody.yield(BlockBody.java:108)
	at org.jruby.runtime.Block.yield(Block.java:184)
	at org.jruby.RubyArray.each(RubyArray.java:1809)
	at org.jruby.RubyArray$INVOKER$i$0$0$each.call(RubyArray$INVOKER$i$0$0$each.gen)
	at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodZeroBlock.call(JavaMethod.java:555)
	at org.jruby.RubyClass.finvokeWithRefinements(RubyClass.java:514)
	at org.jruby.RubyClass.finvoke(RubyClass.java:502)
	at org.jruby.runtime.Helpers.invoke(Helpers.java:444)
	at org.jruby.RubyBasicObject.callMethod(RubyBasicObject.java:393)
	at org.jruby.RubyEnumerator.__each__(RubyEnumerator.java:400)
	at org.jruby.RubyEnumerator.each(RubyEnumerator.java:396)
	at org.jruby.RubyEnumerator.each(RubyEnumerator.java:414)
	at org.jruby.RubyEnumerator$INVOKER$i$each.call(RubyEnumerator$INVOKER$i$each.gen)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:84)
	at org.jruby.runtime.callsite.CachingCallSite.callIter(CachingCallSite.java:93)
	at uri_3a_classloader_3a_.jruby.kernel.enumerator.invokeOther0:each(uri:classloader:/jruby/kernel/enumerator.rb:127)
	at uri_3a_classloader_3a_.jruby.kernel.enumerator.RUBY$block$make_proc$2(uri:classloader:/jruby/kernel/enumerator.rb:127)
	at java.lang.invoke.DirectMethodHandle$Holder.invokeStatic(java.base@13.0.2/DirectMethodHandle$Holder)
	at java.lang.invoke.LambdaForm$MH/0x00000008005b3c40.invoke(java.base@13.0.2/LambdaForm$MH)
	at java.lang.invoke.Invokers$Holder.invokeExact_MT(java.base@13.0.2/Invokers$Holder)
	at org.jruby.runtime.CompiledIRBlockBody.yieldDirect(CompiledIRBlockBody.java:148)
	at org.jruby.runtime.BlockBody.yield(BlockBody.java:106)
	at org.jruby.runtime.Block.yield(Block.java:184)
	at org.jruby.exceptions.CatchThrow.enter(CatchThrow.java:32)
	at org.jruby.RubyKernel.rbCatch(RubyKernel.java:1189)
	at org.jruby.RubyKernel$INVOKER$s$rbCatch.call(RubyKernel$INVOKER$s$rbCatch.gen)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:182)
	at org.jruby.runtime.callsite.CachingCallSite.callIter(CachingCallSite.java:191)
	at uri_3a_classloader_3a_.jruby.kernel.enumerator.invokeOther1:catch(uri:classloader:/jruby/kernel/enumerator.rb:126)
	at uri_3a_classloader_3a_.jruby.kernel.enumerator.RUBY$block$make_proc$1(uri:classloader:/jruby/kernel/enumerator.rb:126)
	at java.lang.invoke.DirectMethodHandle$Holder.invokeStatic(java.base@13.0.2/DirectMethodHandle$Holder)
	at java.lang.invoke.LambdaForm$MH/0x00000008005b3840.invoke(java.base@13.0.2/LambdaForm$MH)
	at java.lang.invoke.Invokers$Holder.invokeExact_MT(java.base@13.0.2/Invokers$Holder)
	at org.jruby.runtime.CompiledIRBlockBody.callDirect(CompiledIRBlockBody.java:138)
	at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58)
	at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:52)
	at org.jruby.runtime.Block.call(Block.java:139)
	at org.jruby.RubyProc.call(RubyProc.java:318)
	at org.jruby.RubyGenerator.each(RubyGenerator.java:104)
	at org.jruby.RubyGenerator$INVOKER$i$0$0$each.call(RubyGenerator$INVOKER$i$0$0$each.gen)
	at org.jruby.RubyClass.finvokeWithRefinements(RubyClass.java:514)
	at org.jruby.RubyBasicObject.send(RubyBasicObject.java:1748)
	at org.jruby.RubyBasicObject$INVOKER$i$send.call(RubyBasicObject$INVOKER$i$send.gen)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:84)
	at org.jruby.runtime.callsite.CachingCallSite.callIter(CachingCallSite.java:93)
	at uri_3a_classloader_3a_.jruby.kernel.enumerator.invokeOther3:__send__(uri:classloader:/jruby/kernel/enumerator.rb:64)
	at uri_3a_classloader_3a_.jruby.kernel.enumerator.RUBY$block$initialize$1(uri:classloader:/jruby/kernel/enumerator.rb:64)
	at java.lang.invoke.DirectMethodHandle$Holder.invokeStatic(java.base@13.0.2/DirectMethodHandle$Holder)
	at java.lang.invoke.LambdaForm$MH/0x00000008005b3c40.invoke(java.base@13.0.2/LambdaForm$MH)
	at java.lang.invoke.Invokers$Holder.invokeExact_MT(java.base@13.0.2/Invokers$Holder)
	at org.jruby.runtime.CompiledIRBlockBody.yieldDirect(CompiledIRBlockBody.java:148)
	at org.jruby.runtime.IRBlockBody.yieldSpecific(IRBlockBody.java:66)
	at org.jruby.runtime.Block.yieldSpecific(Block.java:153)
	at org.jruby.ext.fiber.ThreadFiber.lambda$createThread$0(ThreadFiber.java:300)
	at org.jruby.ext.fiber.ThreadFiber$$Lambda$1555/0x0000000801abb840.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@13.0.2/ThreadPoolExecutor.java:1128)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@13.0.2/ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(java.base@13.0.2/Thread.java:830)
```

Please tell me if I can help more. Thanks!

**Comment by headius** [comment]
This output looks right. I can see that this is stuck on another Lazy enumerator at our internal enumerator.rb:229:

https://github.com/jruby/jruby/blob/e4402ab6a8e95888a294f4c5ef280d99341e3fd2/core/src/main/ruby/jruby/kernel/enumerator.rb#L226-L232

I see I've not really explained the problem here so I'll do that for posterity.

All non-CRuby implementations of Ruby currently in use use native threads to emulate fibers. Where on CRuby, a single native thread can multiplex between all the fibers created on that thread, we must use native threads because the JVM does not allow stack-swapping in this way. The problem here is that native threads on the JVM act as "GC roots", keeping objects in memory that are referenced anywhere on that thread's native stack. Compare this to CRuby, where the fiber is just a ball of data pulled off of and restored onto the thread as needed.

JRuby's fiber has had numerous tweaks to avoid leaking fibers, but one area is a bit harder than others: the Lazy enumerator. When using our current pure-Ruby Lazy with a fiber, we end up with the fiber object ending up on its own native stack, which means that to the GC it is forever a live reference and it will not go away.

A non-Ruby implementation of Lazy might avoid this, but I have not yet attempted to write one.

**Comment by boris-petrov** [comment]
@headius - thanks a lot for your detailed explanation and support!

A couple of questions:

1) Will this be somehow fixed/helped by project Loom? Not that it matters that much, just curious.

2) Why wasn't this a problem in older versions of JRuby? Can't an older implementation of this be returned?

How come no one is hitting this? I would guess that many people use JRuby with Capybara which obviously uses lazy enumerators. So this must be affecting a lot of projects.

3) If you think this is "unfixable" in JRuby in the foreseeable future, should we request help from the Capybara people? As I mentioned, in our tests we're hitting 3000+ blocked threads. That won't scale. :smile: They've already committed JRuby-specific code, I think exactly for enumerators, so probably they would again.

Thank you again for your time!

**Comment by Unknown-Guy** [comment]
There was problem previously, but with release of JRuby 9.2.8.0 they changed condition with https://github.com/teamcapybara/capybara/commit/2e052792a59d0c183ff42009cc73bf7e9eae43c6

So it is "broken" since with [Capybara 3.29.0](https://github.com/teamcapybara/capybara/blob/master/History.md#version-3290)

**Comment by headius** [comment]
@boris-petrov 

> Will this be somehow fixed/helped by project Loom? Not that it matters that much, just curious.

Yes, once we can use Loom, we can support "real" fibers that will not end up rooting themselves.

> Why wasn't this a problem in older versions of JRuby? Can't an older implementation of this be returned?

JRuby has always shipped this pure-Ruby Lazy implementation, with some tweaks over the years to at least *reduce* the occurrence of leaks. Of course, it may not be possible to make a fully "safe" version of Lazy, but that's still an open question.

> How come no one is hitting this? I would guess that many people use JRuby with Capybara which obviously uses lazy enumerators. So this must be affecting a lot of projects.

> If you think this is "unfixable" in JRuby in the foreseeable future, should we request help from the Capybara people?

I'm not sure. It's possible that the Lazy use was added more recently, or the use of Lazy + Enumerator#next may have been added.

From @Unknown-Guy it sounds like Capy has changed this code to not use the same pattern anymore. We appreciate this change very much! However we will continue to look for ways to help fix this pattern in the future.

**Comment by boris-petrov** [comment]
@headius - I believe you misunderstood what @Unknown-Guy said - Capybara actually did the *opposite* - they used to work-around JRuby's troubles before and they removed the work-around a couple of versions ago. So now it's broken.

I'll open an issue there requesting help. Thanks for your time!

**Comment by headius** [comment]
@boris-petrov Yeah you're right, I read that diff wrong. They were using non-lazy enumeration on JRuby and then switched it on once our Lazy impl started working better. So we still have an issue here on versions after that commit.

**Comment by headius** [comment]
@boris-petrov Oh there's one other thing I forgot to ask for... if you can reproduce again, try using VisualVM (download the "standalone" version [here](https://visualvm.github.io/)) and do the following:

* Connect to the JRuby instance (it should show up in the list)
* Go to the Threads tab and capture a Thread Dump from there. Save that output somewhere.
* Go to the Monitor tab and capture a Heap Dump. Save that somewhere.
* Now do the same thing but try hitting the "Perform GC" button until the number of fiber threads in the Threads tab seems to go down.

What I'm looking for here is to try to eliminate any dangling fibers that *should* get cleaned up, so we can see only the ones that *won't* be cleaned up.

**Comment by headius** [comment]
Oh and if you can zip up that output and put it somewhere for me, I'd appreciate it :-)

**Comment by boris-petrov** [comment]
@headius - I did what you requested. The thread-count and the heap-dump size didn't move one bit even when I performed GC 30 times. The threads' stacks are the same as the one above. I ran fewer tests this time so the dumps are smaller I guess but still big - the zipped heap dump is 125 MB. The zipped thread dump is small so I sent you an email with it. Please tell me if you need the heap dump and where to upload it.

**Comment by headius** [comment]
@boris-petrov Thanks for doing that. I have this in my queue and will try to look into it again this week.

**Comment by boris-petrov** [comment]
@headius - thanks! Note, however, that at least my use-case has been "fixed" - Capybara has a new version that works fine with JRuby. I can't speak for the other people in this thread but as far as I'm concerned - this issue is (very) low priority and you could spend your time on something else. :)

**Comment by headius** [comment]
Punting this to 9.3.1 for reexamination along with #6773. Our handling of orphaned fibers may need to be more aggressive, skipping ensure clauses to match CRuby behavior and to avoid ensures re-vivifying the fiber and leading to fiber-shutdown races.
